src/jenkins.py

```diff
@@ -705,3 +708,196 @@ def get_agent_name(unit_name: str) -> str:
+                _get_plugin_name(dependency) for dependency in dependencies.split(", ")
+            )
+        except ValidationError as exc:
+            logger.error("Invalid plugin dependency, %s", exc)
+            continue
+
+    return dependency_lookup
+
+
+def _set_jenkins_system_message(message: str, container: ops.Container) -> None:
+    """Set a system message on Jenkins.
+
+    Args:
+        message: The system message to display.
+        container: The Jenkins workload container.
+    """
+    jcasc_yaml = container.pull(JCASC_CONFIG_FILE_PATH, encoding="utf-8").read()
+    config = yaml.safe_load(jcasc_yaml)
+    config["jenkins"]["systemMessage"] = message
+    container.push(
+        JCASC_CONFIG_FILE_PATH, yaml.dump(config), encoding="utf-8", user=USER, group=GROUP
+    )
+
+
+def _traverse_dependencies(
+    plugin: str, dependency_lookup: typing.Mapping[str, typing.Iterable[str]], seen: set[str]
+) -> typing.Iterable[str]:
+    """Traverse through plugin dependencies that have not yet been traversed recursively.
+
+    Args:
+        plugin: The plugin to recurse through the dependencies.
+        dependency_lookup: The plugin and it's dependencies lookup table.
+        seen: a set indicating whether a plugin has already been traversed.
+
+    Yields:
+        Plugin and it's dependents.
+    """
+    if plugin in seen or plugin not in dependency_lookup:
+        return
+    yield plugin
+    seen.add(plugin)
+    for dependency in dependency_lookup.get(plugin, []):
+        yield from _traverse_dependencies(dependency, dependency_lookup, seen)
+
+
+def _get_allowed_plugins(
+    top_level_plugins: typing.Iterable[str],
+    dependency_lookup: typing.Mapping[str, typing.Iterable[str]],
+) -> typing.Iterable[str]:
+    """Get the plugin short names of allowed plugins and its dependencies.
+
+    Args:
+        top_level_plugins: The allowed plugins short names to add to allowed plugins with its
+            dependencies.
+        dependency_lookup: The plugin dependency lookup table.
+
+    Yields:
+        The allowed plugin short name.
+    """
+    seen: set[str] = set()
+    for plugin in top_level_plugins:
+        if plugin in seen:
+            continue
+        yield from _traverse_dependencies(plugin, dependency_lookup, seen)
+
+
+def _get_top_level_plugins(
+    plugins: typing.Iterable[str], dependency_lookup: typing.Mapping[str, typing.Iterable[str]]
+) -> typing.Iterable[str]:
+    """Get top level plugins that are not dependencies to other plugins.
+
+    Args:
+        plugins: Plugins to extract top level plugins from.
+        dependency_lookup: The dependency lookup table.
+
+    Returns:
+        All plugins that are not dependency to another plugin.
+    """
+    dependent_plugins: set[str] = set()
+    for _, dependencies in dependency_lookup.items():
+        dependent_plugins = dependent_plugins.union(dependencies)
+
+    return set(plugins) - dependent_plugins
+
+
+def remove_unlisted_plugins(
```

Comment:
I'm not seeing any call where the `client` argument is set. Perhaps you'd want to remove it

Reply:
I'm not sure I fully understand what you mean,
the client is initialized in the following code
`client = client if client is not None else _get_client(get_admin_credentials(container))`
and is used to run groovy script, deleting plugins and restarting jenkins.
Would this be what you meant?

Reply:
No, what I mean is that the function signature defined a `client` argument that isn't passed in any function call

Reply:
Oh, the client is initialized inside the code but for test cases, the client is initialized outside as a mock and is injected. Hence the client in the parameters. :)

Reply:
Wouldn't it lead to a simpler code to extract that client into a property and mock it? It ill also simplify the logic in L686